### ISSUE TITLE
request for keyboard support

### ISSUE THREAD

**Issue by sgdoerfler** [rpkamp/chrome-minimal-bookmarks-tree#42]
Any chance of keyboard support? It would be great if we could use arrow keys and Enter to select a menu entry or open and close folders, as in Firefox on Windows, and other browsers. Ideally you should be able to press an alphanumeric key to move to the next entry starting with that letter too (and open it, if unique).

**Comment by rpkamp** [comment]
Should also delete a bookmark when the "delete" key is pressed.

Keyboard support must be a toggle in the settings that is disabled by default, as it's pretty much a power user tool. When enabled the first folder/bookmark visible must be selected when opening the extension. Also, a keyboard shortcut to open the extension must be set - if that's not possible it doesn't really make sense to support keyboards anyway.

**Comment by oleedd** [comment]
> When enabled the first folder/bookmark visible must be selected when opening the extension.

Better to do as it is implemented in Windows: when opening - nothing is selected but after pressing "⇩" the second visible item should be selected.

> Also, a keyboard shortcut to open the extension must be set

Hm. It is nice but I don't think Chrome will allow to do that for security purposes.

**Comment by oleedd** [comment]
https://www.maketecheasier.com/assign-custom-hotkeys-chrome-extensions/

**Comment by rpkamp** [comment]
Yes, I saw that, but it seems that link is gone in newer versions of chrome :slightly_frowning_face: 

It seems however that it's possible to suggest a shortcut in the extension manifest: https://developer.chrome.com/apps/commands :slightly_smiling_face:

**Comment by sgdoerfler** [comment]
You can set a shortcut for an extension that acts like clicking on its icon by visiting chrome://extensions/shortcuts. (Or you can navigate there by selecting More tools.../Extensions from Chrome's main hamburger menu, then clicking on the hamburger menu in the upper left and selecting Keyboard Shortcuts.)

Setting a shortcut lets you display the top-level menus of this extension. But of course it won't let you use the keyboard to then select a particular menu entry.

**Comment by oleedd** [comment]
I checked and `chrome.commands` really works but at least two keys should be used together and one should be Ctrl or Alt. Is it worth of it? Faster to click the icon.

**Comment by oleedd** [comment]
I checked - `chrome.commands` really works and opens the popup but at least two keys should be pressed together (not possible with only one key) and one of them should be Ctrl or Alt. Is it worth to open the extension by pressing two buttons (almost all letters are reserved by Chrome)? Faster to click the icon.

**Comment by rpkamp** [comment]
Hitting two keys is a lot faster than first grabbing the mouse and then moving it to the icon.

Also, you don't have use it if you don't want to, the mouse will still work as well.

**Comment by oleedd** [comment]
Good. Then you can start to implement this issue. It is quite easy, I think.
About arrows - I think it is important that at first no item selected (to make nice using both the arrows and the mouse) but after pressing `⇩⇧` the previous/next item is selected. Also need auto scrolling.

**Comment by rpkamp** [comment]
Since I hardly use the extension anymore myself, and keyboard support really is a nice to have, rather than a must have it's _really_ low on my priority list. I might implement it at some point, but maybe not all.

If someone really wants to see this done, I'd be happy to accept a Pull Request :slightly_smiling_face:

**Comment by oleedd** [comment]
BTW I really need only the delete button. I can write the solution for that if you want.
Moreover, maybe better to make two options - for the delete key and for the arrows. Hotkeys to open the extension can't be disabled because they are set in the manifest.

**Comment by oleedd** [comment]
BTW I really need only the delete key. I can write the solution for that if you want.
Moreover, maybe better to make two options - for the delete key and for the arrows. Hotkeys to open the extension can't be disabled because they are set in the manifest.

**Comment by oleedd** [comment]
BTW I really need only the delete key. I can write a solution for that if you want.
Moreover, maybe better to make two options - for the delete key and for the arrows. Hotkeys to open the extension can't be disabled because they are set in the manifest.

**Comment by rpkamp** [comment]
Just the delete key won't be that hard. I might look into that later.

**Comment by oleedd** [comment]
Cool. It is really very useful if need to delete many retired or unnecessary bookmarks.

**Comment by oleedd** [comment]
See a realization of the "Delete" key:
(added before `bm.addEventListener('mousedown', function (event) {`)
```javascript
document.addEventListener("keydown", function(e) {
	if (e.key == "Delete") {
		chrome.bookmarks.remove(selectedBookmark.dataset.itemId, function () {
          		selectedBookmark.remove();
		});
	}
})
```
(added after `d = document.createElement('li');`)
```javascript
d.addEventListener("mouseenter", function() {
	window.selectedBookmark = this;
})
```

**Comment by rpkamp** [comment]
That would break the context menu because `selectedBookmark` will keep changing as you move the mouse.

**Comment by oleedd** [comment]
I can't reproduce any breaking.
It deletes a bookmark under the mouse. It doesn't depend on the right click menu and the menu will keep working.

**Comment by oleedd** [comment]
I can't reproduce any breaking.
It deletes a bookmark under the mouse. It doesn't depend on the right click menu and the menu keeps working.

**Comment by oleedd** [comment]
New the best possible solution:

(added after before `bm.addEventListener('mousedown', function (event) {`)
```javascript
addEventListener("keydown", function(e) {
	if (e.key == "Delete") {
		var elem = document.elementFromPoint(moveEvent.clientX, moveEvent.clientY).parentNode;
		chrome.bookmarks.remove(elem.dataset.itemId, function () {
          		elem.remove();
		});
	}
})
```
(added after `return function pointCB(event) {`)
```javascript
window.moveEvent = event;
```
Can you check this?

**Comment by oleedd** [comment]
Where are you?

**Comment by oleedd** [comment]
New the best possible solution:

(added after before `bm.addEventListener('mousedown', function (event) {`)
```javascript
addEventListener("keydown", function(e) {
	if (e.key == "Delete") {
		var elem = document.elementFromPoint(moveEvent.clientX, moveEvent.clientY).parentNode;
		chrome.bookmarks.remove(elem.dataset.itemId, function () {
          		elem.remove();
		});
	}
})
```
(added after `return function pointCB(event) {`)
```javascript
window.moveEvent = event;
```
So you can add this because it really works.

**Comment by oleedd** [comment]
I wait for reply.

**Comment by rpkamp** [comment]
I've just added support for the "Delete" key in the new version that is rolling out now (5.2.0).
It should arrive in the next hour or so.
You do need to enable the option "Keyboard support" in the options, which for now only supports the delete key.
